juniper_official/Linecard/check-pfe-discards.rule

```diff
@@ -0,0 +1,321 @@
+/*
+ * Detects PFE discards and notifies when anomalies are found.
+ * One inputs control detection
+ * 
+ *   1) discard-count, is the threshold that causes the rule to report
+ *      an anomaly.  By default it's 1. This rule will set a dashboard
+ *      color to red when *all* the discards are greater than
+ *      'discard-count' for 60s. otherwise color is set to green.
+ */  
+healthbot {
+    topic linecard.pfe {
+        rule check-pfe-discards {
+             /*
+             * Monitors the pfe discards count on PFE. Notifies via the dashboard
+             * colors when the flaps increase by {{discard-count}} for every collected
+             * metric during a 60s time range (red). Otherwise the color is set to green
+             * 
+             * Use FPC and PFE names as key for rule.
+             */  
+            keys [ fpc-no pfe-no ];
+            synopsis "Packet forwarding engine hardware discard statistics analyzer";
+            description "Collects packet forwarding engine hardware discard statistics  and notifies when discard count increases";
+            sensor pfe-sensor {
+                synopsis "iAgent netconf sensor definition";
+                description "Collects data from network device netconf rpc using get-pfe-traffic-statistics";
+                iAgent {
+                    file pfe-discard-statistics.yml;
+                    table PfeDiscardStatisticsTable;
+                    frequency 60s;
+                }
+            }
+            field drop-threshold {
+                constant {
+                    value "{{discard-count}}";
+                }
+                type integer;
+                description "Drop count increase threshold";
+            }
+            field fpc-no {
+                sensor pfe-sensor {
+                    path pfe-fpc-index;
+                }
+                type string;
+                description "FPC slot number";
+            }
+            field pfe-no {
+                sensor pfe-sensor {
+                    path pfe-index;
+                }
+                type string;
+                description "PFE number";
+            }
+            trigger data-error-discard {
+                synopsis "PFE data error discards KPI";
+                description "Sets health based on the change in PFE data error discard count";
+                frequency 60s;
+                term is-data-error-discard {
+                    when {
+                        increasing-at-least-by-value "$data-error-discard" {
```

Comment:
Shouldn't we be using a time-range here to ensure that we deal with jitter/delays in iAgent and writing to Influx?

Reply:
Discards should be evaluated every time(sensor frequency) we receives data as these are critical hence time-range is not required.

Reply:
Hi Venkat,
 The issue is that both sensor and trigger frequency are set to 60s. That means that for the lookback window for checking for data is 120s - which doesn't necessarily guarantee the presence of two data points. Ideally this should be 3x.

Reply:
As we discussed increased trigger frequency to 75s